framework/src/media/audio/resample/samplerate.c

```diff
@@ -0,0 +1,550 @@
+		uint32_t part = FRACPART_VALUE(fp_index);
+		int32_t k;
+		for (k = 0; k < channels_num; k++) {
+			int32_t s1 = input[(whole / quoti) * channels_num + k];
+			int32_t s2 = input[((whole + 1) / quoti) * channels_num + k];
+			*output++ = clip(CALC_NEW_SAMPLE(s1, s2, part));
+		}
+	}
+}
+
+/**
+ * @brief   It handles samplerate down scaling in ratio 1.* with fraction.
+ * @remarks ratio 1.*, e.g. 24K->16K, 16K->11.025K, ...
+ * @param   src: pointer to resampler object.
+ * @param   num_frames_out: number of frames will be stored in output buffer.
+ * @param   frac: ratio of orig samplerate being divided by target samplerate.
+ * @return  void
+ * @see
+ */
+static void downresample_frac(resampler_t *src, int32_t num_frames_out, float frac)
+{
+	const int16_t *input = src->in_buffer;
+	int16_t *output = src->out_buffer;
+	int32_t channels_num = src->channels_num;
+	float step_float = frac;
+	uint32_t step = TO_16_16_FIXED(step_float);
+	uint32_t fp_index = 0;   // 16.16 fixed point value
+
+	int32_t j;
+	for (j = 0; j < num_frames_out; ++j, fp_index += step) {
+		uint32_t whole = INTPART_VALUE(fp_index);
+		uint32_t part = FRACPART_VALUE(fp_index);
+		int32_t k;
+		for (k = 0; k < channels_num; k++) {
+			int32_t s1 = input[whole * channels_num + k];
+			int32_t s2 = input[(whole + 1) * channels_num + k];
+			*output++ = clip(CALC_NEW_SAMPLE(s1, s2, part));
+		}
+	}
+}
+
+/**
+ * @brief   It handles samplerate down scaling in ratio 2.0/3.0 without fraction.
+ * @remarks ratio 2.0/3.0, e.g. 48K->16K, 48K->24K, 44.1K->22.05K...
+ * @param   src: pointer to resampler object.
+ * @param   num_frames_in: number of frames in above input buffer.
+ * @param   quoti: quotient of orig samplerate being divided by target samplerate, it is the ratio mentioned above.
+ * @return  void
+ * @see     downresample_2_1_filter_frac()
+ */
+static void downresample_int(resampler_t *src, int32_t num_frames_in, int32_t quoti)
+{
+	const int16_t *input = src->in_buffer;
+	int16_t *output = src->out_buffer;
+	int32_t channels_num = src->channels_num;
+	int32_t num_samples_in = num_frames_in * channels_num;
+
+	int32_t i;
+	for (i = 0; i < num_samples_in; i += quoti) {
+		output[i / quoti] = clip(fir_convolve(input + i, filter_22khz_coeff, NUM_COEFF_22KHZ, channels_num));
+	}
+}
+
+/**
+ * @brief   It handles samplerate down scaling in special ratio.
+ * @remarks The special ratio is 1.*, e.g. 44.1K->32K, 22.05K->16K, 11.025K->8K
+ * @param   src: pointer to resampler object.
+ * @param   num_frames_in: number of frames in above input buffer.
+ * @param   num_frames_out: number of frames will be stored in output buffer.
+ * @return  void
+ * @see
+ */
+static void downresample_441_320(resampler_t *src, int32_t num_frames_in, int32_t num_frames_out)
+{
+	int16_t *input = src->in_buffer;
+	int16_t *output = src->out_buffer;
+	int32_t channels_num = src->channels_num;
+	int32_t num_samples_in = num_frames_in * channels_num;
+	float step_float = (float)441 / (float)320;
```

Comment:
This is specific case, 441 to 320,
why don't you define ?
#define DOWN_RESAMPLE_441_320_RATIO XXXXX